include/wil/com.h

```diff
@@ -3002,6 +3002,135 @@ namespace wil
 
 #endif // __IObjectWithSite_INTERFACE_DEFINED__
 
+
+// if C++17 or greater
+#if WIL_HAS_CXX_17
+#ifdef WIL_ENABLE_EXCEPTIONS
+namespace details
+{
+    template<typename>
+    struct com_enumerator_next_traits;
+
+    template<typename Itf, typename T, typename Ret>
+    struct com_enumerator_next_traits<Ret(__stdcall Itf::*)(ULONG, T*, ULONG*)>
+    {
+        using Interface = Itf;
+        using Result = T;
+    };
+
+    template<typename Itf, typename T, typename Ret>
+    struct com_enumerator_next_traits<Ret(__stdcall Itf::*)(ULONG, T*, ULONG*) noexcept>
+    {
+        using Interface = Itf;
+        using Result = T;
+    };
+
+
+    template<typename Interface>
+    struct com_enumerator_traits
+    {
+        using Result = typename com_enumerator_next_traits<decltype(&Interface::Next)>::Result;
+        // If the result is a COM pointer type (IFoo*), then we use wil::com_ptr<IFoo>. Otherwise, we use IFoo*.
+        using smart_result = std::conditional_t<std::is_pointer_v<Result> && std::is_base_of_v<::IUnknown, std::remove_pointer_t<Result>>,
+            wil::com_ptr<std::remove_pointer_t<Result>>, Result>;
+    };
+}
+
+template <typename IEnumType, typename TStoredType = typename details::com_enumerator_traits<IEnumType>::smart_result>
+struct iterator

```

Comment:
```suggestion
struct com_enum_iterator
```
Given that this is directly in the `wil` namespace, the name should be more specific